sktime/markov/msm/augmented_msm.py

```diff
@@ -0,0 +1,623 @@
+        # (changes are smaller than 1e-3).
+        # These do not always converge together, but usually within a few steps of each other.
+        # A better heuristic for the latter may be necessary. For realistic cases (the two ubiquitin examples in [1])
+        # this yielded results very similar to those with more stringent convergence criteria
+        # (changes smaller than 1e-9) with convergence times
+        # which are seconds instead of tens of minutes.
+        #
+
+        converged = False  # Convergence flag for lagrange multipliers
+        i = 0
+        die = False
+        while i <= self.maxiter:
+            pi_hat_old = state.pi_hat.copy()
+            state.update_pi_hat()
+            if not np.all(state.pi_hat > 0):
+                state.pi_hat = pi_hat_old.copy()
+                die = True
+                self._log.warning("pihat does not have a finite probability for all states, terminating")
+            state.update_m_hat()
+            state.update_Q()
+
+            if i > 1:
+                X_old = np.copy(state.X)
+                state.update_X_and_pi()
+                if np.any(state.X[nonzero_counts] < 0) and i > 0:
+                    die = True
+                    self._log.warning(
+                        "Warning: new X is not proportional to C... reverting to previous step and terminating")
+                    state.X = X_old
+
+            if not converged:
+                self._newton_lagrange(state, count_matrix)
+            else:  # once Lagrange multipliers are converged compute likelihood here
+                transition_matrix = state.X / state.pi[:, None]
+                _ll_new = state.log_likelihood_biased(count_matrix, transition_matrix)
+                state.log_likelihoods.append(_ll_new)
+
+            # General case fixed-point iteration
+            if len(count_outside) > 0:
+                if i > 1 and np.all((np.abs(state.d_m_hat) / self.uncertainties) < self.convergence_criterion_lagrange) \
+                        and not converged:
+                    self._log.info("Converged Lagrange multipliers after %i steps..." % i)
+                    converged = True
+            # Special case
+            else:
+                if np.abs(state.log_likelihoods[-2] - state.log_likelihoods[-1]) < 1e-8:
+                    self._log.info("Converged Lagrange multipliers after %i steps..." % i)
+                    converged = True
+            # if Lagrange multipliers are converged, check whether log-likelihood has converged
+            if converged and np.abs(state.log_likelihoods[-2] - state.log_likelihoods[-1]) < 1e-8:
+                self._log.info("Converged pihat after %i steps..." % i)
+                die = True
+            if die:
+                break
+            if i == self.maxiter:
+                self._log.info("Failed to converge within %i iterations. "
+                               "Consider increasing max_iter(now=%i)" % (i, self.max_iter))
```

Comment:
same question w/ format string. also would it be helpful to print convergence ratio in error message for error diagnosis purposes